src/widgets/forms.js

```diff
@@ -103,6 +102,15 @@ field[ns.ui('Form').uri] = field[ns.ui('Group').uri] =
       for (const x in already) already2[x] = 1
       already2[key] = 1
       const formDoc = form.doc ? form.doc() : null // @@ if blank no way to know
+      const weight0 = kb.any(form, ui('weight'), null, formDoc) // Say 0-3
```

Comment:
**issue**: For me as a reviewer not directly involved in these changes it is very hard to crasp what all the changes in this file (form.js) are all about. The whole file is nearly 2.000 lines of code, It is even not obvious in which functions the changes are done. It is an anomymous function here, with no name, what is it's purpose? Corresponding tests could serve as specification of the desired behaviour, but there are none. Imho the file needs a whole lot of refactoring and testing before adding new features, otherwise we will be adding a whole lot of bugs as well.

Reply:
I agree that forms.js needs dramatic restructuring, splitting it into main files, conversing to TS at the same time, and extracting named functions to reduce the complexity. As has been done with `forms/basic` for example, but applied to  the other fields.  Not on this branch.  (And look out for file-wide variable renaming we may want to do before we split it up.)

Reply:
I agree, we should not try to solve all problems at the same time. That is not what I suggest. However I highly recommend to add at least tests that cover the changes made to such a file. In my experience this is a viable strategy to get started with testing and the situation can improve continuously from there instead of waiting for the big future day where everything is refactored and tests are added.

Reply:
I have added tests for the autocomplete form field, with two paths to 'accept' and one cancel path.   But only testing the case of there being existing data, not when there is no existing data, or testing all combinations of parameters to the form.